execution/executor/src/lib.rs

```diff
@@ -341,38 +364,47 @@ where
                 vm_output.gas_used(),
```

Comment:
I wonder if it's possible to get rid of the state_tree field from TransactionData - this would (a) reduce the total work (as we wouldn't have to create all these intermediate chained trees), and (b) simplify the bulk update (replacing lines 299-312) logic and make it easier to parallelize. If this is possible, I'd suggest that the bulk update returns the final tree after the block and the intermediate hashes. 

Sorry if this is stupid - but I hope it's not. I looked at TransactionData and the only two methods accessing it's state_tree member are the root_hash() - which we would still have (would just store root_hash in TransactionData instead of state_tree shared pointer), and prune_state_tree(). I grep'ed that the only place where prune_state_tree() is called is when the block is committed, in which case the state trees of all transactions in the block are pruned in order. If we just had a single final block state_tree and pruned that, I believe the result would be the same?

Reply:
cc @lightmark @msmouse

Reply:
from what I understand, the per transaction state tree is to support state queries on every version (instead of every block)

Reply:
@zekun000 made the core point here. The reason is we need per-txn root hashes according to the current sys design. Otherwise I agree.

Reply:
1. the root hash is part of the `TransactionInfo` which authenticates EACH txn along with its outputs (events, writeset, gas_used). Without it, you can prove things like your account balance at a specific version.

2. The execution of a txn can depend on outputs of previous txns in the same block, which is not yet committed. Such state is accessed via `VerifiedStateView.speculative_state` which is a reference to the `SparseMerkleTree` (search for the only non-test usage of SparseMerkleTree::get() to see it)

^ To remove 1 we need alternative way to authenticate the account state at each version (like the "on-chain writeset" proposal, where a merkle summary of the write set is part of the txn_info) and/or remove such ability (you can only get authenticated state result at block boundaries, like in the "blocks as first class data structure" discussion, where the top level accumulator is one of blocks, not txns.)

To remove 2 as a road blocker to execution parallelization, we need the executor (with the help of VM/txn parser) to be smart about the state dependencies between txns. I'm aware you guys are already on it.

^ hopefully this can eliminate the need for a meeting :D

Reply:
I think for 1), we might not need the entire merkle tree node in `TransactionData`, only a root hash would sufficient so that we. don't need to change anything in storage.
 
For 2), it seems that we only need the last tree because there couldn't be a branch from middle of a transaction block right?

Reply:
@zekun000 @lightmark - indeed, returning intermediate root hashes shouldn't be too hard after @runtian-zhou's refactoring. Trying to understand whether returning the intermediate (inside the block) stitched trees is. To me it seemed that those were only stored in TransactionData and used either to get a hash (which we can still have) or by a prune() method, but this method was called in commit for all trees in succession which seemed also easy to replace. I can't find the call to the get() yet in question for (2) @msmouse mentioned - if this is called for one of the SparseMerkleTrees after a whole block, then as @runtian-zhou said, we are totally cool - we still would have stitched together SparseMerkleTree roots but only after blocks.

Reply:
@runtian re 1: the root hash is a result on top of all the intermediate nodes calculated in the speculative state tree; if you mean the root hash can be calculated in the next stage, that's true. But we are again going towards the "split thing to sequential steps at high level and parallelize every tiny step with multithreading" route which I think is simply "wrong" IMHO :D -- again we should try to overlap things at high level before everything else, which tends to yield less overhead and more consistent/balanced/smooth workload.  re 2: I don't think you are talking about the same thing with me.. 

@gelash  Sorry I didn't mean a literal search. https://github.com/diem/diem/blob/07c6fe5cf6f0ddfde7dcffbe485a58499bd552c0/storage/storage-interface/src/state_view.rs#L152 (and you can really use an IDE if you prefer the fancy route like me.) To access the in mem state previous txns yielded yes you can introduce a cache for those only and without calculating intermediate nodes right now, but see my point above -- I'm worried that you guys are breaking things to tiny parallelize-able pieces everywhere and introduce overhead.

Reply:
Alden thanks for the time and responses. I don't think we explained what the plan was after this change, but knowing it I strongly disagree with the "split thing to sequential steps at high level and parallelize every tiny step with multithreading" judgement. I may be wrong, in which case I'd love to know, but here is why I think the way I do:
1. We didn't intend to split anything - in fact, right now updates are added one-at-a-time, and unnecessary intermediate data-structures are returned (when we suspect returning just intermediate root hashes may be sufficient). If we are right (and that's what we are trying to find out), then the current solution is over-engineered, so we could improve it and that could also help performance without parallelizing anything.
2. We would add bulk-update API to the SparseMerkleTree library, and in it, we wouldn't parallelize every tiny thing, we could just parallelize the first 2 or 4 recursive calls down the tree. Since most of the time is spent in hashing, and recursion for all updates should traverse down quite a bit, we hope to get a solid ~2-4x with a couple lines change on top of the bulk-update API.
3. We wouldn't be getting rid of any existing functionalities from the SparseMerkleTree library, instead providing a new functionality to in fact not have to sequentially do tiny steps of adding things, if we know ahead of time we are adding a sequence of things and intermediate root hashes could suffice.

We discussed the fact that once the big projects happen (e.g. execution & consensus), the way these data-structures are used could change dramatically and what we are doing may not matter as much, but the current benchmarking shows that process_write_set is quite heavy, and also affects execution benchmark in terms of seeing what the limits are for the folks doing vm parallelization experiments. So if it's possible and we can land 2 diffs (including this one) in short time-frame that gets rid of unnecessary return values and improves performance, I don't see why we shouldn't go for it.

Thanks for the code reference - I grep'ed the wrong thing, should've just followed from the file (don't want to trade emacs for fancy IDE's just yet :). Following the call chain, it's provided with the SparseMerkleTrees after blocks, never in between transactions of some block, so it seems to me we should be able to keep this behavior easily.

Reply:
@gelash 
(K. I see you are on the king of fanciness path.)

What you said makes sense. I didn't realize the VM manages the cache inside of a block internally already.
Without regression and too much added complexity I'm okay with possible changes.